client/session.go

```diff
@@ -1177,6 +1177,81 @@ func (s *session) FetchBootstrapBlocksFromPeers(
 	return result.result, nil
 }
 
+func (s *session) FetchRepairBlocksFromPeers(
+	namespace ts.ID,
+	shard uint32,
+	repairBlocks []block.ReplicaMetadata,
+	opts result.Options,
+) (PeerBlocksIter, error) {
+	var (
+		complete = int64(0)
+		doneCh   = make(chan error, 1)
+		outputCh = make(chan peerBlocksDatapoint, 4096)
+		result   = newStreamBlocksResult(s.opts, opts, s.multiReaderIteratorPool, outputCh)
+		onDone   = func(err error) {
+			atomic.StoreInt64(&complete, 1)
+			select {
+			case doneCh <- err:
+			default:
+			}
+		}
+		m = s.streamFromPeersMetricsForShard(shard)
+	)
+
+	peers, err := s.peersForShard(shard)
+	if err != nil {
+		return nil, err
+	}
+	peersByHost := make(map[string]peer, len(peers))
+	for _, peer := range peers {
+		peersByHost[peer.Host().ID()] = peer
+	}
+
+	go func() {
+		for atomic.LoadInt64(&complete) == 0 {
+			m.fetchBlocksFromPeers.Update(1)
+			time.Sleep(gaugeReportInterval)
+		}
+		m.fetchBlocksFromPeers.Update(0)
+	}()
+
+	// Transform provided repairBlocks into channel
+	metadataCh := make(chan blocksMetadata, 4096)
+	go func() {
+		for _, rb := range repairBlocks {
+			peer, ok := peersByHost[rb.Peer.ID()]
+			if !ok {
+				err := fmt.Errorf("unable to find peer: %v", rb.Peer.ID())
+				onDone(err) // bail early
```

Comment:
I had it this way to sanitise input, didn't consider the topology change case at all. That said, the way I've got the repair code structured, we merge units of peer block replicas individually anyway. So yeah, logging a warning and continuing makes sense.